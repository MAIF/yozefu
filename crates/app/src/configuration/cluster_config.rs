//! module defining the configuration structure of the application

use indexmap::IndexMap;
use resolve_path::PathResolveExt;
use serde::{Deserialize, Serialize};
use std::{collections::HashMap, path::PathBuf};
use url::Url;

use super::Configuration;

/// List of kafka properties that are a file location.
const KAFKA_PROPERTIES_WITH_LOCATIONS: [&str; 6] = [
    "ssl.ca.location",
    "ssl.certificate.location",
    "ssl.key.location",
    "ssl.keystore.location",
    "ssl.crl.location",
    "ssl.engine.location",
];

impl Default for ClusterConfig {
    fn default() -> Self {
        Self {
            url_template: Some(default_url_template()),
            schema_registry: None,
            kafka: Default::default(),
        }
    }
}

/// Specific configuration for a cluster
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Clone)]
pub struct ClusterConfig {
    /// A placeholder url that will be used when you want to open a kafka record in the browser
    pub url_template: Option<String>,
    /// Schema registry configuration
    #[serde(skip_serializing_if = "Option::is_none")]
    pub schema_registry: Option<SchemaRegistryConfig>,
    // Kafka consumer properties for this cluster, see <https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md> for more details
    pub kafka: IndexMap<String, String>,
}

impl ClusterConfig {
    /// Returns the kafka properties for the given cluster.
    pub fn normalize_paths(self) -> Self {
        let mut cloned = self.clone();
        for key in KAFKA_PROPERTIES_WITH_LOCATIONS {
            if let Some(path) = cloned.kafka.get(key) {
                let normalized_path = PathBuf::from(path)
                    .resolve()
                    .canonicalize()
                    .map(|d| d.display().to_string())
                    .unwrap_or(path.to_string());
                cloned.kafka.insert(key.to_string(), normalized_path);
            }
        }
        cloned
    }
}

/// Schema registry configuration of a given cluster
#[derive(Debug, Deserialize, PartialEq, Eq, Serialize, Clone)]
pub struct SchemaRegistryConfig {
    /// Url of the schema registry
    pub url: Url,
    /// HTTP headers to be used when communicating with the schema registry
    #[serde(default = "HashMap::default")]
    pub headers: HashMap<String, String>,
}

fn default_url_template() -> String {
    "http://localhost/cluster/{topic}/{partition}/{offset}".to_string()
}

impl Configuration for ClusterConfig {
    fn kafka_config_map(&self) -> HashMap<String, String> {
        let mut properties = HashMap::new();
        properties.extend(self.kafka.clone());
        properties
    }
}
