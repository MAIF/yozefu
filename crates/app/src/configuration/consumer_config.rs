use serde::{Deserialize, Serialize};

/// Configuration for the kafka consumer
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Clone)]
#[cfg_attr(test, derive(schemars::JsonSchema))]
pub struct ConsumerConfig {
    pub buffer_capacity: usize,
    pub timeout_in_ms: u64,
}

impl Default for ConsumerConfig {
    fn default() -> Self {
        Self {
            buffer_capacity: 1000,
            timeout_in_ms: 10,
        }
    }
}
