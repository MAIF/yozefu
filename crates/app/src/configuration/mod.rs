use std::collections::HashMap;

use lib::Error;
use rdkafka::{ClientConfig, config::FromClientConfig};

mod cluster_config;
mod consumer_config;
mod global_config;
mod internal_config;
mod yozefu_config;

pub use cluster_config::ClusterConfig;
pub use cluster_config::KAFKA_PROPERTIES_WITH_LOCATIONS;
pub use cluster_config::SchemaRegistryConfig;
pub use consumer_config::ConsumerConfig;
pub use global_config::GlobalConfig;
pub use internal_config::InternalConfig;
pub use yozefu_config::YozefuConfig;

pub trait Configuration {
    /// Returns the kafka properties
    fn kafka_config_map(&self) -> HashMap<String, String>;

    /// Properties you can set for the kafka consumer: <https://github.com/confluentinc/librdkafka/blob/master/CONFIGURATION.md>
    fn create_kafka_consumer<T>(&self) -> Result<T, Error>
    where
        T: FromClientConfig,
    {
        Self::kafka_client_config_from_properties(self.kafka_config_map().clone())
            .create()
            .map_err(|e| e.into())
    }

    fn kafka_client_config_from_properties(
        kafka_properties: HashMap<String, String>,
    ) -> ClientConfig {
        let mut config = ClientConfig::new();
        config.set_log_level(rdkafka::config::RDKafkaLogLevel::Emerg);
        for (key, value) in kafka_properties {
            config.set(key, value);
        }

        config
    }
}
