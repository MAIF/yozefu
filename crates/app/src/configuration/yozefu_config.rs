//! module defining the configuration of the yozefu application

use super::{Configuration, SchemaRegistryConfig};
use crate::{configuration::ClusterConfig, APPLICATION_NAME};
use std::{collections::HashMap, path::PathBuf};

/// composed of kafka properties and
/// an optional user-specific configuration.
#[derive(Debug, Clone)]
pub struct YozefuConfig {
    cluster_config: ClusterConfig,
    pub logs_file: Option<PathBuf>,
    pub export_directory: Option<PathBuf>,
}

impl YozefuConfig {
    pub fn new(cluster_config: ClusterConfig) -> Self {
        Self {
            cluster_config: cluster_config.normalize_paths(),
            logs_file: None,
            export_directory: None,
        }
    }

    pub fn url_template(&self) -> Option<String> {
        self.cluster_config.url_template.clone()
    }

    pub fn schema_registry(&self) -> Option<SchemaRegistryConfig> {
        self.cluster_config.schema_registry.clone()
    }

    pub fn with_exported_directory(self, exported_directory: PathBuf) -> Self {
        Self {
            cluster_config: self.cluster_config,
            logs_file: self.logs_file,
            export_directory: Some(exported_directory),
        }
    }

    pub fn with_logs_file(self, logs_file: PathBuf) -> Self {
        Self {
            cluster_config: self.cluster_config,
            logs_file: Some(logs_file),
            export_directory: self.export_directory,
        }
    }

    pub fn set_kafka_property(&mut self, key: &str, value: &str) {
        self.cluster_config
            .kafka
            .insert(key.to_string(), value.to_string());
    }

    /// Overrides the kafka properties with the properties provided by the user
    pub fn update_kafka_properties(self, kafka_properties: HashMap<String, String>) -> Self {
        Self {
            cluster_config: ClusterConfig {
                url_template: None,
                schema_registry: None,
                kafka: indexmap::IndexMap::from_iter(kafka_properties),
            },
            logs_file: self.logs_file,
            export_directory: self.export_directory,
        }
    }
}

impl Configuration for YozefuConfig {
    /// Returns the kafka properties
    fn kafka_config_map(&self) -> HashMap<String, String> {
        let mut config_map = self.cluster_config.kafka_config_map();

        // Default properties
        for (key, value) in [
            ("group.id", APPLICATION_NAME),
            ("enable.auto.commit", "false"),
        ] {
            if !config_map.contains_key(key) {
                config_map.insert(key.into(), value.into());
            }
        }
        config_map
    }
}
