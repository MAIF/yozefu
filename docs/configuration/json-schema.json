{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "GlobalConfig",
  "description": "Configuration of the application",
  "type": "object",
  "properties": {
    "default_url_template": {
      "description": "A placeholder url that will be used when you want to open a kafka record in the browser",
      "type": "string",
      "default": "http://localhost/cluster/{topic}/{partition}/{offset}"
    },
    "initial_query": {
      "description": "The initial search query when you start the UI",
      "type": "string"
    },
    "theme": {
      "type": "string",
      "default": "light"
    },
    "clusters": {
      "description": "The kafka properties for each cluster",
      "type": "object",
      "additionalProperties": {
        "$ref": "#/$defs/ClusterConfig"
      }
    },
    "default_kafka_config": {
      "description": "The default kafka properties inherited for every cluster",
      "type": "object",
      "additionalProperties": {
        "type": "string"
      }
    },
    "history": {
      "description": "History of past search queries",
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "show_shortcuts": {
      "description": "Show shortcuts",
      "type": "boolean",
      "default": true
    },
    "export_directory": {
      "type": "string",
      "default": "./yozefu-exports"
    }
  },
  "required": [
    "initial_query",
    "clusters",
    "default_kafka_config",
    "history"
  ],
  "$defs": {
    "ClusterConfig": {
      "description": "Specific configuration for a cluster",
      "type": "object",
      "properties": {
        "url_template": {
          "description": "A placeholder url that will be used when you want to open a kafka record in the browser",
          "type": [
            "string",
            "null"
          ]
        },
        "schema_registry": {
          "description": "Schema registry configuration",
          "anyOf": [
            {
              "$ref": "#/$defs/SchemaRegistryConfig"
            },
            {
              "type": "null"
            }
          ]
        },
        "kafka": {
          "type": "object",
          "additionalProperties": {
            "type": "string"
          }
        }
      },
      "required": [
        "kafka"
      ]
    },
    "SchemaRegistryConfig": {
      "description": "Schema registry configuration of a given cluster",
      "type": "object",
      "properties": {
        "url": {
          "description": "Url of the schema registry",
          "type": "string",
          "format": "uri"
        },
        "headers": {
          "description": "HTTP headers to be used when communicating with the schema registry",
          "type": "object",
          "additionalProperties": {
            "type": "string"
          },
          "default": {}
        }
      },
      "required": [
        "url"
      ]
    }
  }
}