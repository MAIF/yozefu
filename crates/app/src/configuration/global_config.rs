//! module defining the configuration structure of the application

use std::{
    collections::HashMap,
    fs,
    path::{Path, PathBuf},
};

use directories::ProjectDirs;
use indexmap::IndexMap;
use itertools::Itertools;
use lib::Error;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::{APPLICATION_NAME, configuration::ConsumerConfig};

use super::cluster_config::{ClusterConfig, SchemaRegistryConfig};

const EXAMPLE_PROMPTS: &[&str] = &[
    r#"timestamp between "2 hours ago" and "1 hour ago" limit 100 from beginning"#,
    r#"offset > 100000 and value contains "music" limit 10"#,
    r#"key == "ABC" and timestamp >= "2 days ago""#,
];

/// Configuration of the application
#[derive(Debug, Deserialize, Serialize, PartialEq, Eq, Clone)]
#[cfg_attr(test, derive(schemars::JsonSchema))]
pub struct GlobalConfig {
    /// Path of this config
    #[serde(skip)]
    pub path: PathBuf,
    /// Path to the Yozefu directory containing themes, config, filters...
    #[serde(skip)]
    pub yozefu_directory: PathBuf,
    /// The file to write logs to
    #[serde(skip)]
    pub logs: Option<PathBuf>,
    /// A placeholder url that will be used when you want to open a kafka record in the browser
    #[serde(default = "default_url_template")]
    pub default_url_template: String,
    /// The initial search query when you start the UI
    pub initial_query: String,
    /// The theme to use in the TUI
    #[serde(default = "default_theme")]
    pub theme: String,
    /// The theme to use for syntax highlighting
    pub highlighter_theme: Option<String>,
    /// The kafka properties for each cluster
    pub clusters: IndexMap<String, ClusterConfig>,
    #[serde(default)]
    /// The default configuration for the yozefu kafka consumer
    pub consumer: ConsumerConfig,
    /// The default kafka properties inherited for every cluster
    pub default_kafka_config: IndexMap<String, String>,
    /// History of past search queries
    pub history: Vec<String>,
    /// Show shortcuts
    #[serde(default = "default_show_shortcuts")]
    pub show_shortcuts: bool,
    #[serde(default = "default_export_directory")]
    pub export_directory: PathBuf,
}

fn default_url_template() -> String {
    "http://localhost/cluster/{topic}/{partition}/{offset}".to_string()
}

fn default_export_directory() -> PathBuf {
    PathBuf::from(format!("./{APPLICATION_NAME}-exports"))
}

fn default_theme() -> String {
    if cfg!(target_os = "windows") {
        "dark".to_string()
    } else {
        "light".to_string()
    }
}

fn default_show_shortcuts() -> bool {
    true
}

impl TryFrom<&PathBuf> for GlobalConfig {
    type Error = Error;

    fn try_from(path: &PathBuf) -> Result<Self, Self::Error> {
        Ok(Self {
            path: path.clone(),
            yozefu_directory: Self::yozefu_directory()?,
            logs: None,
            default_url_template: default_url_template(),
            history: EXAMPLE_PROMPTS
                .iter()
                .map(|e| (*e).to_string())
                .collect_vec(),
            initial_query: "from end - 10".to_string(),
            clusters: IndexMap::default(),
            default_kafka_config: IndexMap::default(),
            theme: default_theme(),
            highlighter_theme: None,
            show_shortcuts: true,
            export_directory: default_export_directory(),
            consumer: ConsumerConfig::default(),
        })
    }
}

impl GlobalConfig {
    /// The default config file path
    pub fn path() -> Result<PathBuf, Error> {
        Self::yozefu_directory().map(|d| d.join("config.json"))
    }

    /// The default yozefu directory containing themes, filters, config...
    pub fn yozefu_directory() -> Result<PathBuf, Error> {
        ProjectDirs::from("io", "maif", APPLICATION_NAME)
            .ok_or(Error::Error(
                "Failed to find the yozefu configuration directory".to_string(),
            ))
            .map(|e| e.config_dir().to_path_buf())
    }

    /// Reads a configuration file.
    pub fn read(file: &Path) -> Result<Self, Error> {
        let content = fs::read_to_string(file);
        if let Err(e) = &content {
            return Err(Error::Error(format!(
                "Failed to read the configuration file {:?}: {}",
                file.display(),
                e
            )));
        }

        let content = content.unwrap();
        let mut config: Self = serde_json::from_str(&content).map_err(|e| {
            Error::Error(format!(
                "Failed to parse the configuration file {:?}: {}",
                file.display(),
                e
            ))
        })?;
        config.yozefu_directory = Self::yozefu_directory()?;
        config.path = file.to_path_buf();
        Ok(config)
    }

    /// Returns the name of the logs file
    pub fn logs_file(&self) -> PathBuf {
        self.logs
            .clone()
            .unwrap_or(self.path.parent().unwrap().join("application.log"))
    }

    /// Returns the name of the logs file
    pub fn themes_file(&self) -> PathBuf {
        self.yozefu_directory.join("themes.json")
    }

    /// Returns the list of available theme names.
    pub fn themes(&self) -> Vec<String> {
        let file = self.themes_file();
        let content = fs::read_to_string(file).unwrap_or("{}".to_string());
        let themes: HashMap<String, Value> = serde_json::from_str(&content).unwrap_or_default();
        themes
            .keys()
            .map(std::string::ToString::to_string)
            .collect_vec()
    }

    /// Returns the name of the directory containing wasm filters
    pub fn filters_dir(&self) -> PathBuf {
        let dir = self.yozefu_directory.join("filters");
        let _ = fs::create_dir_all(&dir);
        dir
    }

    /// web URL template for a given cluster
    pub fn url_template_of(&self, cluster: &str) -> String {
        self.clusters
            .get(cluster)
            .and_then(|e| e.url_template.clone())
            .unwrap_or(self.default_url_template.clone())
    }

    /// Consumer config of a given cluster
    pub(crate) fn consumer_config_of(&self, cluster: &str) -> ConsumerConfig {
        self.clusters
            .get(cluster)
            .and_then(|e| e.consumer.clone())
            .unwrap_or(self.consumer.clone())
    }

    /// Returns the schema registry configuration for the given cluster.
    pub fn schema_registry_config_of(&self, cluster: &str) -> Option<SchemaRegistryConfig> {
        self.clusters
            .get(cluster.trim())
            .and_then(|config| config.schema_registry.clone())
    }
}

#[test]
fn generate_json_schema_for_global_config() {
    use schemars::schema_for;
    let mut schema = schema_for!(GlobalConfig);
    schema.insert("$id".into(), "https://raw.githubusercontent.com/MAIF/yozefu/refs/heads/main/docs/json-schemas/global-config.json".into());
    fs::write(
        PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .parent()
            .unwrap()
            .join("docs")
            .join("json-schemas")
            .join("global-config.json"),
        serde_json::to_string_pretty(&schema).unwrap(),
    )
    .unwrap();
}
