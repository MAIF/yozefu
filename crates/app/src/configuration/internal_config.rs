//! module defining the configuration of the yozefu application

use std::{collections::HashMap, path::PathBuf};

use chrono::Local;

use crate::configuration::{GlobalConfig, SchemaRegistryConfig};

use super::{Configuration, yozefu_config::YozefuConfig};

#[derive(Debug, Clone)]
pub struct InternalConfig {
    specific: YozefuConfig,
    pub global: GlobalConfig,
    output_file: PathBuf,
}

impl Configuration for InternalConfig {
    fn kafka_config_map(&self) -> HashMap<String, String> {
        self.specific.kafka_config_map()
    }
}

impl InternalConfig {
    pub fn new(specific: YozefuConfig, global: GlobalConfig) -> Self {
        let directory = match &specific.export_directory {
            Some(e) => e,
            None => &global.export_directory,
        }
        .clone();

        let output_file = directory.join(format!(
            "export-{}.json",
            // Windows does not support ':' in filenames
            Local::now()
                .to_rfc3339_opts(chrono::SecondsFormat::Secs, false)
                .replace(':', "-"),
        ));

        Self {
            specific,
            global,
            output_file,
        }
    }

    /// web URL template for a given cluster
    pub fn url_template_of(&self, cluster: &str) -> String {
        match &self.specific.url_template() {
            Some(url) => url.to_string(),
            None => self.global.url_template_of(cluster),
        }
    }

    /// Returns the schema registry configuration for the given cluster.
    pub fn schema_registry_config_of(&self, cluster: &str) -> Option<SchemaRegistryConfig> {
        match &self.specific.schema_registry() {
            Some(schema_registry) => Some(schema_registry.clone()),
            None => self.global.schema_registry_config_of(cluster),
        }
    }

    /// Returns the output file path for exported kafka records.
    pub fn output_file(&self) -> &PathBuf {
        &self.output_file
    }
}
