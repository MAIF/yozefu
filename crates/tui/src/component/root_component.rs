//! This component is handles the main layout of the TUI
//! and renders components based on the current context.
use app::configuration::GlobalConfig;
use copypasta::{ClipboardContext, ClipboardProvider};
use std::{
    collections::HashMap,
    sync::{Arc, Mutex},
};
use tracing::warn;

use crossterm::event::{KeyCode, KeyEvent, KeyModifiers};
use ratatui::{
    Frame,
    layout::{Constraint, Direction, Flex, Layout, Margin, Rect},
    widgets::{Clear, Paragraph},
};
use tokio::sync::{mpsc::UnboundedSender, watch::Receiver};

use crate::{Action, Notification, action::Level, error::TuiError, records_buffer::BufferAction};

use super::{
    Component, ComponentName, ConcurrentRecordsBuffer, State, footer_component::FooterComponent,
    header_component::HeaderComponent, help_component::HelpComponent,
    progress_bar_component::ProgressBarComponent, record_details_component::RecordDetailsComponent,
    records_component::RecordsComponent, schemas_component::SchemasComponent,
    search_component::SearchComponent, topic_details_component::TopicDetailsComponent,
    topics_and_records_component::TopicsAndRecordsComponent, topics_component::TopicsComponent,
};

pub(crate) struct RootComponent {
    components: HashMap<ComponentName, Arc<Mutex<dyn Component>>>,
    views: Vec<ComponentName>,
    state: State,
    focus_history: Vec<ComponentName>,
    focus_order: Vec<ComponentName>,
    progress_bar: ProgressBarComponent,
    buffer_rx: Receiver<BufferAction>,
    action_tx: Option<UnboundedSender<Action>>,
}

impl RootComponent {
    #[allow(clippy::arc_with_non_send_sync)]
    pub fn new(
        query: String,
        selected_topics: Vec<String>,
        config: &GlobalConfig,
        records: &'static ConcurrentRecordsBuffer,
        state: State,
    ) -> Self {
        let buffer_rx = records.lock().map(|e| e.channels.clone().1).ok().unwrap();
        let mut footer = FooterComponent::default();
        footer.show_shortcuts(config.show_shortcuts);

        let topics_component = Arc::new(Mutex::new(TopicsComponent::new(selected_topics)));

        let mut components: [Arc<Mutex<dyn Component>>; 10] = [
            topics_component.clone(),
            Arc::new(Mutex::new(RecordsComponent::new(records))),
            Arc::new(Mutex::new(TopicDetailsComponent::default())),
            Arc::new(Mutex::new(RecordDetailsComponent::new(&state))),
            Arc::new(Mutex::new(SearchComponent::new(
                &query,
                config.history.clone(),
                config.filters_dir(),
            ))),
            Arc::new(Mutex::new(footer)),
            Arc::new(Mutex::new(HelpComponent::default())),
            Arc::new(Mutex::new(SchemasComponent::new())),
            Arc::new(Mutex::new(HeaderComponent::default())),
            Arc::new(Mutex::new(FooterComponent::default())),
        ];

        components[components.len() - 1] = Arc::new(Mutex::new(TopicsAndRecordsComponent::new(
            topics_component.clone(),
            components[1].clone(),
        )));

        let components: HashMap<ComponentName, Arc<Mutex<dyn Component>>> = components
            .into_iter()
            .map(|c| {
                let id = c.lock().unwrap().id();
                (id, c)
            })
            .collect();
        Self {
            components,
            buffer_rx,
            progress_bar: ProgressBarComponent::new(400),
            views: vec![ComponentName::TopicsAndRecords],
            focus_order: focus_order_of(&ComponentName::TopicsAndRecords),
            focus_history: vec![],
            state,
            action_tx: Default::default(),
        }
    }

    fn focus_next(&mut self, current: &ComponentName) -> ComponentName {
        let index = self
            .focus_order
            .iter()
            .position(|e| e == current)
            .unwrap_or(self.focus_order.len() - 1);
        match index == self.focus_order.len() - 1 {
            true => self.focus_order.first().unwrap().clone(),
            false => self.focus_order[index + 1].clone(),
        }
    }

    fn focus(&mut self, current: ComponentName) -> Result<(), TuiError> {
        self.state.focused = current;
        let mut shortcuts = self
            .components
            .get(&self.state.focused)
            .unwrap()
            .lock()
            .unwrap()
            .shortcuts();

        if self.state.focused == ComponentName::RecordDetails {
            shortcuts.extend(
                self.components
                    .get(&ComponentName::Records)
                    .unwrap()
                    .lock()
                    .unwrap()
                    .shortcuts(),
            );
        }

        self.action_tx
            .as_ref()
            .unwrap()
            .send(Action::Shortcuts(shortcuts, self.views.len() == 1))?;
        Ok(())
    }

    fn close(&mut self) {
        self.views.pop();
        if self.views.is_empty() {
            self.action_tx.as_ref().unwrap().send(Action::Quit).unwrap();
        } else {
            self.focus_order = focus_order_of(self.views.last().unwrap());

            let last_focused_component = self
                .focus_history
                .pop()
                .unwrap_or(self.focus_order.first().unwrap().clone());

            let focus = match self.focus_order.contains(&self.state.focused) {
                true => self.state.focused.clone(),
                false => last_focused_component,
            };
            self.focus(focus).unwrap();
        }
        self.notify_footer().unwrap();
    }

    fn toggle_view(&mut self, view: ComponentName) -> Result<(), TuiError> {
        if self.views.last().unwrap() == &view {
            self.close();
        } else {
            self.action_tx
                .as_ref()
                .unwrap()
                .send(Action::NewView(view))?;
        }
        Ok(())
    }

    fn focus_previous(&mut self, current: &ComponentName) -> ComponentName {
        let index = self
            .focus_order
            .iter()
            .position(|e| e == current)
            .unwrap_or(1);
        match index == 0 {
            true => self.focus_order.last().unwrap().clone(),
            false => self.focus_order[index - 1].clone(),
        }
    }

    fn notify_footer(&self) -> Result<(), TuiError> {
        if !self.views.is_empty() {
            self.action_tx.as_ref().unwrap().send(Action::ViewStack((
                self.views.first().unwrap().clone(),
                self.focus_history.clone(),
            )))?;
        }
        Ok(())
    }
}

impl Component for RootComponent {
    fn register_action_handler(&mut self, tx: UnboundedSender<Action>) {
        self.action_tx = Some(tx.clone());
        for component in self.components.values_mut() {
            component
                .lock()
                .unwrap()
                .register_action_handler(tx.clone());
        }
    }

    fn id(&self) -> ComponentName {
        ComponentName::Main
    }

    fn init(&mut self) -> Result<(), TuiError> {
        self.notify_footer()?;
        self.action_tx.as_ref().unwrap().send(Action::Shortcuts(
            self.components
                .get(&self.state.focused)
                .unwrap()
                .lock()
                .unwrap()
                .shortcuts(),
            true,
        ))?;
        Ok(())
    }

    fn handle_key_events(&mut self, key: KeyEvent) -> Result<Option<Action>, TuiError> {
        match key.code {
            KeyCode::Tab => {
                let new_focus = self.focus_next(&self.state.focused.clone());
                self.focus(new_focus)?;
            }
            KeyCode::BackTab => {
                let new_focus = self.focus_previous(&self.state.focused.clone());
                self.focus(new_focus)?;
            }
            KeyCode::Char('r') if key.modifiers.contains(KeyModifiers::CONTROL) => {
                self.action_tx.as_ref().unwrap().send(Action::Refresh)?;
            }
            KeyCode::Char('f') if key.modifiers.contains(KeyModifiers::CONTROL) => {
                self.state.focused = ComponentName::Search;
                return Ok(None);
            }
            KeyCode::Char('/') | KeyCode::Char(':')
                if self.state.focused != ComponentName::Search =>
            {
                self.state.focused = ComponentName::Search;
                return Ok(None);
            }
            KeyCode::Char('c') if key.modifiers.contains(KeyModifiers::CONTROL) => {
                self.action_tx.as_ref().unwrap().send(Action::Quit)?;
            }
            KeyCode::Char('o')
                if key.modifiers.contains(KeyModifiers::CONTROL) && self.views.len() == 1 =>
            {
                match self.views.first().unwrap() {
                    ComponentName::Records => self.views[0] = ComponentName::TopicsAndRecords,
                    ComponentName::TopicsAndRecords => self.views[0] = ComponentName::Records,
                    _ => warn!(
                        "View '{}' does not support toggling. This instruction should not be unreachable",
                        self.views.first().unwrap()
                    ),
                }
                self.notify_footer()?;
                if self.views.len() == 1 {
                    self.focus_order = focus_order_of(&self.views[0]);
                    self.state.focused = match self.focus_order.contains(&self.state.focused) {
                        true => self.state.focused.clone(),
                        false => self.focus_order.first().unwrap().clone(),
                    };
                    self.action_tx
                        .as_ref()
                        .unwrap()
                        .send(Action::RefreshShortcuts)?;
                }
                return Ok(None);
            }
            KeyCode::Char('h') if key.modifiers.contains(KeyModifiers::CONTROL) => {
                self.toggle_view(ComponentName::Help)?;
                return Ok(None);
            }
            KeyCode::Esc => self.close(),
            _ => (),
        };
        let focused_component = self.components.get(&self.state.focused).unwrap();
        focused_component.lock().unwrap().handle_key_events(key)?;
        if self.state.focused == ComponentName::RecordDetails
            && (key.code == KeyCode::Up || key.code == KeyCode::Down)
        {
            self.components
                .get(&ComponentName::Records)
                .unwrap()
                .lock()
                .unwrap()
                .handle_key_events(key)?;
        }
        Ok(None)
    }

    fn update(&mut self, action: Action) -> Result<Option<Action>, TuiError> {
        match action {
            Action::RefreshShortcuts => {
                let focused = self.state.focused.clone();
                self.focus(focused)?;
            }
            Action::NewView(ref action) => {
                if action == &ComponentName::RecordDetails
                    && self.state.focused == ComponentName::RecordDetails
                {
                    return Ok(None);
                }
                self.focus_history.push(self.state.focused.clone());
                self.focus_history.dedup();
                let last_focused = self.focus_history.last().unwrap().clone();
                self.focus_order = focus_order_of(action);
                self.views = self
                    .views
                    .iter()
                    .filter(|a| a != &action)
                    .cloned()
                    .collect();

                self.views.push(action.clone());
                match self.focus_order.contains(&last_focused) {
                    true => self.focus(last_focused),
                    false => self.focus(
                        self.focus_order
                            .first()
                            .unwrap_or_else(|| {
                                panic!("I think you forgot to define focus order of '{action}'")
                            })
                            .clone(),
                    ),
                }?;
                self.notify_footer()?;
            }
            Action::RecordsToRead(length) => {
                self.progress_bar.set_length(length);
                let mut a = self.buffer_rx.clone();
                let BufferAction::Stats(mut stats) = *a.borrow_and_update();
                stats.total_to_read = length;
            }
            Action::CopyToClipboard(ref content) => {
                let mut ctx = ClipboardContext::new().unwrap();
                self.action_tx
                    .as_ref()
                    .unwrap()
                    .send(Action::Notification(Notification::new(
                        Level::Info,
                        "Copied to clipboard".to_string(),
                    )))?;
                ctx.set_contents(content.to_string()).unwrap();
            }
            _ => (),
        }
        for component in self.components.values_mut() {
            component.lock().unwrap().update(action.clone())?;
        }
        Ok(None)
    }

    fn draw(&mut self, f: &mut Frame<'_>, rect: Rect, _: &State) -> Result<(), TuiError> {
        if rect.width < 20 && rect.height < 4 {
            let [area] = Layout::horizontal([Constraint::Length(4)])
                .flex(Flex::Center)
                .areas(rect);
            let [area] = Layout::vertical([Constraint::Length(1)])
                .flex(Flex::Center)
                .areas(area);

            f.render_widget(
                Paragraph::new(
                    String::from_utf8(vec![74, 32, 226, 157, 164, 239, 184, 143]).unwrap(),
                ),
                area,
            );
            return Ok(());
        }
        let mut a = self.buffer_rx.clone();
        let BufferAction::Stats(stats) = *a.borrow_and_update();
        self.progress_bar.set_progress(stats.read);
        f.render_widget(Clear, rect);

        let last_view = self.views.last().unwrap();
        let main_component = self
            .components
            .get(last_view)
            .unwrap_or_else(|| panic!("Unable to find component '{last_view}'"));

        let chunks: std::rc::Rc<[Rect]> = Layout::default()
            .direction(Direction::Vertical)
            .constraints([
                Constraint::Min(2),
                Constraint::Percentage(100),
                Constraint::Min(3),
                Constraint::Min(1),
            ])
            .split(rect.inner(Margin::new(1, 1)));

        main_component
            .lock()
            .unwrap()
            .draw(f, chunks[1], &self.state)?;
        self.components
            .get_mut(&ComponentName::Header)
            .unwrap()
            .lock()
            .unwrap()
            .draw(f, chunks[0], &self.state)?;
        self.components
            .get_mut(&ComponentName::Search)
            .unwrap()
            .lock()
            .unwrap()
            .draw(f, chunks[2], &self.state)?;
        self.components
            .get_mut(&ComponentName::Footer)
            .unwrap()
            .lock()
            .unwrap()
            .draw(f, chunks[3].inner(Margin::new(1, 0)), &self.state)?;

        //f.render_widget(self.progress_bar.clone(), rect);
        //f.render_widget(search_block, chunks[1]);

        Ok(())
    }
}

pub fn focus_order_of(component: &ComponentName) -> Vec<ComponentName> {
    match component {
        ComponentName::RecordDetails => vec![ComponentName::RecordDetails, ComponentName::Search],
        ComponentName::Records => vec![ComponentName::Records, ComponentName::Search],
        ComponentName::Schemas => vec![ComponentName::Schemas, ComponentName::Search],
        ComponentName::TopicsAndRecords => vec![
            ComponentName::Topics,
            ComponentName::Records,
            ComponentName::Search,
        ],
        ComponentName::TopicDetails => vec![ComponentName::TopicDetails, ComponentName::Search],
        ComponentName::Help => vec![ComponentName::Help, ComponentName::Search],
        _ => vec![],
    }
}
