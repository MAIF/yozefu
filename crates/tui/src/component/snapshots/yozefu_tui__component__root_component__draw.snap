---
source: crates/tui/src/component/root_component.rs
expression: terminal.backend()
---
"                                                                                                                        "
"   localhost   Topics                                                                                                  "
"                                                                                                                        "
"  ┏ Topics [2] ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓ ╭ Records ────────────────────────────────────────────╮"
"  ┃ [x] topic1                                                 ┃ │Timestamp                Topic   Offset         Key  │"
"  ┃ [x] topic2                                                 ┃ │                                                     │"
"  ┃                                                            ┃ │1970-01-01T01:00  movi..ers[0]      314  7f12bd3b..  │"
"  ┃                                                            ┃ │                                                     │"
"  ┃                                                            ┃ │                                                     │"
"  ┃                                                            ┃ │                                                     │"
"  ┃                                                            ┃ │                                                     │"
"  ┃                                                            ┃ │                                                     │"
"  ┃                                                            ┃ │                                                     │"
"  ┃                                                            ┃ │                                                     │"
"  ┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛ ╰─────────────────────────────────────────────────────╯"
"  ╭ Search ────────────────────────────────────────────────────────────────────────────────────────────────────────────╮"
"  │ from begin                                                                                                         │"
"  ╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯"
"                                                                                                                        "
"                                                                                                                        "
